src/libsyntax/json.rs

```diff
@@ -330,10 +333,24 @@ impl DiagnosticSpan {
             })
         });
 
+        let file_np = &start.file.normalized_pos;
+        let np_start_diff = match file_np.binary_search_by(
+                            |np| np.pos.cmp(&span.lo())) {
+            Ok(i) => file_np[i].diff,
+            Err(i) if i == 0 => 0,
+            Err(i) => file_np[i-1].diff,
+        };
```

Comment:
Let's move this binary search to the `SourceFile`:

```rust
fn original_relative_byte_pos(&self, pos: BytePos) -> BytePos
```

And lets also move `- self.start_pos` bit in there, as normalized absolute positions do not make sense.